json/src/map.rs

```diff
@@ -0,0 +1,335 @@
+    #[inline]
+    pub fn get<Q: ?Sized>(&self, key: &Q) -> Option<&Value>
+        where String: Borrow<Q>,
+              Q: Ord + Eq + Hash
+    {
+        self.0.get(key)
+    }
+
+    /// Returns true if the map contains a value for the specified key.
+    ///
+    /// The key may be any borrowed form of the map's key type, but the ordering
+    /// on the borrowed form *must* match the ordering on the key type.
+    #[inline]
+    pub fn contains_key<Q: ?Sized>(&self, key: &Q) -> bool
+        where String: Borrow<Q>,
+              Q: Ord + Eq + Hash
+    {
+        self.0.contains_key(key)
+    }
+
+    /// Returns a mutable reference to the value corresponding to the key.
+    ///
+    /// The key may be any borrowed form of the map's key type, but the ordering
+    /// on the borrowed form *must* match the ordering on the key type.
+    #[inline]
+    pub fn get_mut<Q: ?Sized>(&mut self, key: &Q) -> Option<&mut Value>
+        where String: Borrow<Q>,
+              Q: Ord + Eq + Hash
+    {
+        self.0.get_mut(key)
+    }
+
+    /// Inserts a key-value pair into the map.
+    ///
+    /// If the map did not have this key present, `None` is returned.
+    ///
+    /// If the map did have this key present, the value is updated, and the old
+    /// value is returned. The key is not updated, though; this matters for
+    /// types that can be `==` without being identical.
+    #[inline]
+    pub fn insert(&mut self, k: String, v: Value) -> Option<Value> {
+        self.0.insert(k, v)
+    }
+
+    /// Removes a key from the map, returning the value at the key if the key
+    /// was previously in the map.
+    ///
+    /// The key may be any borrowed form of the map's key type, but the ordering
+    /// on the borrowed form *must* match the ordering on the key type.
+    #[inline]
+    pub fn remove<Q: ?Sized>(&mut self, key: &Q) -> Option<Value>
+        where String: Borrow<Q>,
+              Q: Ord + Eq + Hash
+    {
+        self.0.remove(key)
+    }
+
+    /// Returns the number of elements in the map.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns true if the map contains no elements.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Gets an iterator over the entries of the map.
+    #[inline]
+    pub fn iter(&self) -> MapIter {
+        MapIter(self.0.iter())
+    }
+
+    /// Gets a mutable iterator over the entries of the map.
+    #[inline]
+    pub fn iter_mut(&mut self) -> MapIterMut {
+        MapIterMut(self.0.iter_mut())
+    }
+
+    /// Gets an iterator over the keys of the map.
+    #[inline]
+    pub fn keys(&self) -> MapKeys {
+        MapKeys(self.0.keys())
+    }
+
+    /// Gets an iterator over the values of the map.
+    #[inline]
+    pub fn values(&self) -> MapValues {
+        MapValues(self.0.values())
+    }
+}
+
+impl Default for Map<String, Value> {
```

Comment:
The derived impls don't work:

- Ord is required for BTreeMap's Default:

    ```
    error[E0277]: the trait bound `K: core::cmp::Ord` is not satisfied
      --> src/map.rs:15:22
       |
    15 | pub struct Map<K, V>(MapImpl<K, V>);
       |                      ^^^^^^^^^^^^^^ the trait `core::cmp::Ord` is not implemented for `K`
    ```

- Hash and Eq are required for all four of LinkedHashMap's impls:

    ```
    error[E0277]: the trait bound `K: core::cmp::Eq` is not satisfied
      --> src/map.rs:15:22
       |
    15 | pub struct Map<K, V>(MapImpl<K, V>);
       |                      ^^^^^^^^^^^^^^ the trait `core::cmp::Eq` is not implemented for `K`
       |

    error[E0277]: the trait bound `K: core::hash::Hash` is not satisfied
      --> src/map.rs:15:22
       |
    15 | pub struct Map<K, V>(MapImpl<K, V>);
       |                      ^^^^^^^^^^^^^^ the trait `core::hash::Hash` is not implement
    ```

What I did was just implement these for \<K=String, Value=Value>. I found this to be the best balance of:

- now that Map is opaque, allowing us to tweak the internal representation for performance or other reasons; other implementations may have tighter trait bounds than BTreeMap or LinkedHashMap
- keeping the definition of Value explicit and obvious; compare:

    ```rust
    pub enum Value {
        Null,
        Bool(bool),
        Number(Number),
        String(String),
        Array(Vec<Value>),
        Object(Map<String, Value>),
    }
    ```

    vs

    ```rust
    pub enum Value {
        Null,
        Bool(bool),
        Number(Number),
        String(String),
        Array(Vec<Value>),
        Object(Map), // of...?
    }
    ```